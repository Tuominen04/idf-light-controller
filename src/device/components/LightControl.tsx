/**
 * @file LightControl.tsx
 * @description
 * UI helper for rendering the Light Control card, including status, color indicator, and control button.
 * 
 * @remarks
 * - Used to display and interact with the light state of a device.
 * - Button disables and shows loading indicator based on device and UI state.
 */

import { ActivityIndicator, Text, TouchableOpacity, View } from "react-native";
import styles from "../../styles/DeviceControlScreen.styles";

/**
 * Renders the Light Control UI card.
 * 
 * @param lightState - Current light state ('on' or 'off')
 * @param loading - Whether the control is loading
 * @param refresh - Whether the control is refreshing
 * @param isConnected - Whether the device is connected
 * @param otaLoading - Whether OTA update is in progress
 * @param toggleLight - Handler to toggle light state
 * @returns JSX.Element
 */
export const renderLightControl = (
  lightState: string, 
  loading: boolean, 
  refresh: boolean, 
  isConnected: boolean, 
  otaLoading: boolean, 
  toggleLight: () => void
) => (
    <View style={styles.controlCard}>
      {/* Card Title */}
      <Text style={styles.cardTitle}>Light Control</Text>
      <View style={styles.lightControl}>
        <View style={styles.lightStatus}>
          {/* Light status indicator color */}
          <View 
            style={[
              styles.lightIndicator, 
              { backgroundColor: lightState === 'on' ? '#FFC107' : '#666' }
            ]} 
            testID="light-indicator"
          />
          {/* Light state text */}
          <Text style={styles.lightStateText}>
            Light is {lightState.toUpperCase()}
          </Text>
        </View>
        <View style={styles.controlButtons}>
          {/* Control button: disables and shows loader based on state */}
          <TouchableOpacity
            style={[
              styles.controlButton, 
              { opacity: loading || !isConnected ? 0.5 : 1 }
            ]}
            onPress={toggleLight}
            disabled={loading || refresh || !isConnected || otaLoading}
            testID="light-control-button"
          >
            {loading || refresh ? (
              // Show loading indicator if loading or refreshing
              <ActivityIndicator size="small" color="#fff" />
            ) : (
              // Show button text based on light state
              <Text style={styles.controlButtonText}>
                {lightState === 'on' ? 'Turn OFF' : 'Turn ON'}
              </Text>
            )}
          </TouchableOpacity>
        </View>
      </View>
    </View>
  );
