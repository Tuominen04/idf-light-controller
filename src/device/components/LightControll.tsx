import { ActivityIndicator, Text, TouchableOpacity, View } from "react-native";
import styles from "../../styles/DeviceControlScreen.styles";

export const renderLightControl = (lightState: string, loading: boolean, refresh: boolean, isConnected: boolean, toggleLight: () => void) => (
    <View style={styles.controlCard}>
      <Text style={styles.cardTitle}>Light Control</Text>
      <View style={styles.lightControl}>
        <View style={styles.lightStatus}>
          <View style={[styles.lightIndicator, { backgroundColor: lightState === 'on' ? '#FFC107' : '#666' }]} />
          <Text style={styles.lightStateText}>Light is {lightState.toUpperCase()}</Text>
        </View>
        <View style={styles.controlButtons}>
          <TouchableOpacity
            style={[styles.controlButton, { opacity: loading || !isConnected ? 0.5 : 1 }]}
            onPress={toggleLight}
            disabled={loading || refresh || !isConnected}
          >
            {loading || refresh ? (
              <ActivityIndicator size="small" color="#fff" />
            ) : (
              <Text style={styles.controlButtonText}>
                {lightState === 'on' ? 'Turn OFF' : 'Turn ON'}
              </Text>
            )}
          </TouchableOpacity>
        </View>
      </View>
    </View>
  );
