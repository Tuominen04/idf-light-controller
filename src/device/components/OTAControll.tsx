import { ActivityIndicator, Modal, Text, TouchableOpacity, View } from "react-native";
import styles from "../../styles/DeviceControlScreen.styles";
import { OTAProgress } from "../../types";
import { otaUrl } from "../../credentials";

export const renderOTAControl = (
  loading: boolean,
  isConnected: boolean,
  otaProgress: OTAProgress | null,
  setShowOTAModal: (show: boolean) => void
) => {
  console.log("Render OTA Control");
  console.log("Render OTA progress", otaProgress);
  return (
    <View style={styles.controlCard}>
      <Text style={styles.cardTitle}>OTA Updates</Text>
      <Text style={styles.cardDescription}>
        Upload new firmware to your device over WiFi
      </Text>
        
        {otaProgress?.in_progress && (
          <View style={styles.otaProgress}>
            <Text style={styles.otaStatusText}>{otaProgress.status}</Text>
            {otaProgress.progress && (
              <View style={styles.progressBar}>
                <View 
                  style={[styles.progressFill, { width: `${otaProgress.progress}%` }]} 
                />
                <Text style={styles.progressText}>{otaProgress.progress.toFixed(0)}%</Text>
              </View>
            )}
          </View>
        )}
        
        <TouchableOpacity
          style={[styles.secondaryButton, { 
            opacity: loading || !isConnected || (otaProgress?.in_progress) ? 0.5 : 1 
          }]}
          onPress={() => setShowOTAModal(true)}
          disabled={loading || !isConnected || (otaProgress?.in_progress)}
        >
          <Text style={styles.secondaryButtonText}>Start OTA Update</Text>
        </TouchableOpacity>
      </View>

  )
};

export const renderOTAModal = (
  showOTAModal: boolean, 
  setShowOTAModal: (show: boolean) => void, 
  startOTAUpdate: () => void, 
  loading: boolean, 
  refresh: boolean, 
  otaLoading: boolean
) => {
  return (
    <Modal
      visible={showOTAModal}
      animationType="slide"
      transparent={true}
            onRequestClose={() => setShowOTAModal(false)}
        >
            <View style={styles.modalOverlay}>
            <View style={styles.modalContent}>
                <Text style={styles.modalTitle}>OTA Update</Text>
                <Text style={styles.modalDescription}>
                Start ota from URL {otaUrl}
                </Text>          
                <View style={styles.modalButtons}>
                <TouchableOpacity
                    style={[styles.modalButton, styles.cancelButton]}
                    onPress={() => setShowOTAModal(false)}
                >
                    <Text style={styles.cancelButtonText}>Cancel</Text>
                </TouchableOpacity>
                <TouchableOpacity
                    style={[styles.modalButton, styles.confirmButton]}
                    onPress={startOTAUpdate}
                    disabled={loading || refresh || otaLoading}
                >
                    {loading || refresh || otaLoading ? (
                    <ActivityIndicator size="small" color="#fff" />
                    ) : (
                        <Text style={styles.confirmButtonText}>Start Update</Text>
                    )}
                </TouchableOpacity>
                </View>
            </View>
            </View>
        </Modal>
    )
};