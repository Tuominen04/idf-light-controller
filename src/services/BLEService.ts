import { BleManager, Device, State } from 'react-native-ble-plx';
import { PermissionsAndroid, Platform } from 'react-native';
import { BLE_CONSTANTS } from '../utils/constants';
import base64 from 'react-native-base64';

interface BLEDevice {
  id: string;
  name: string;
  rssi: number | null;
}

class BLEServiceClass {
  private manager: BleManager;
  private scanning: boolean = false;
  private connectedDevice: Device | null = null;
  private initialized: boolean = false;

  constructor() {
    this.manager = new BleManager();
  }

  // Initialize BLE and request permissions
  async initialize(): Promise<void> {
    // Check if already initialized
    if (this.initialized) {
      console.log('BLE already initialized');
      return;
    }

    console.log('Initializing BLE...');
    
    // Request permissions on Android
    if (Platform.OS === 'android') {
      const apiLevel = parseInt(Platform.Version.toString(), 10);
      
      if (apiLevel >= 31) {
        // Android 12 and above
        const result = await PermissionsAndroid.requestMultiple([
          PermissionsAndroid.PERMISSIONS.BLUETOOTH_SCAN,
          PermissionsAndroid.PERMISSIONS.BLUETOOTH_CONNECT,
          PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
        ]);
        
        const allGranted = Object.values(result).every(
          permission => permission === PermissionsAndroid.RESULTS.GRANTED
        );
        
        if (!allGranted) {
          throw new Error('Bluetooth permissions not granted');
        }
      } else if (apiLevel >= 23) {
        // Android 6 to 11
        const granted = await PermissionsAndroid.request(
          PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
          {
            title: 'Location Permission',
            message: 'Bluetooth Low Energy needs access to your location',
            buttonNeutral: 'Ask Me Later',
            buttonNegative: 'Cancel',
            buttonPositive: 'OK',
          }
        );
        
        if (granted !== PermissionsAndroid.RESULTS.GRANTED) {
          throw new Error('Location permission not granted');
        }
      }
    }

    // Check if Bluetooth is powered on
    const state = await this.manager.state();
    console.log('Current BLE state:', state);
    
    if (state !== State.PoweredOn) {
      // Wait for Bluetooth to be powered on
      await new Promise<void>((resolve) => {
        const subscription = this.manager.onStateChange((newState) => {
          console.log('BLE state changed to:', newState);
          if (newState === State.PoweredOn) {
            subscription.remove();
            resolve();
          }
        }, true);
      });
    }
    
    console.log('BLE initialized successfully');
    this.initialized = true;
  }

  // Scan for ESP32 devices
  async scanForDevices(
    onDeviceFound: (device: BLEDevice) => void,
    duration: number = 10000
  ): Promise<void> {
    if (this.scanning) {
      console.log('Already scanning');
      return;
    }

    console.log('Starting BLE scan for', BLE_CONSTANTS.DEVICE_NAME_PREFIX);
    this.scanning = true;
    const foundDevices = new Map<string, Device>();

    this.manager.startDeviceScan(
      null, // Scan for all services
      { allowDuplicates: true }, // Allow duplicates to update RSSI
      (error, device) => {
        if (error) {
          console.error('Scan error:', error);
          this.scanning = false;
          return;
        }

        if (device && device.name) {
          // Check if device name starts with our prefix
          if (device.name.startsWith(BLE_CONSTANTS.DEVICE_NAME_PREFIX)) {
            // Update or add device
            foundDevices.set(device.id, device);
            
            console.log(`Found device: ${device.name} (${device.id}) RSSI: ${device.rssi}`);
            
            // Notify callback
            onDeviceFound({
              id: device.id,
              name: device.name,
              rssi: device.rssi,
            });
          }
        }
      }
    );

    // Stop scanning after duration
    setTimeout(() => {
      this.stopScan();
    }, duration);
  }

  // Stop scanning
  stopScan(): void {
    if (this.scanning) {
      console.log('Stopping BLE scan');
      this.manager.stopDeviceScan();
      this.scanning = false;
    }
  }

  // Connect to device
  async connectToDevice(deviceId: string): Promise<Device> {
    try {
      console.log('Connecting to device:', deviceId);
      
      // Stop scanning first
      this.stopScan();
      
      // Connect to device
      const device = await this.manager.connectToDevice(deviceId, {
        autoConnect: false,
        timeout: 15000,
      });
      
      // Discover services and characteristics
      await device.discoverAllServicesAndCharacteristics();
      
      this.connectedDevice = device;
      console.log('Connected to device:', device.name);
      
      return device;
    } catch (error) {
      console.error('Connection error:', error);
      throw error;
    }
  }

  // Send WiFi credentials (uncomment when needed)
  async sendWiFiCredentials(ssid: string, password: string): Promise<void> {
    if (!this.connectedDevice) {
      throw new Error('No device connected');
    }

    const credentials = { ssid, password };
    const credentialsJson = JSON.stringify(credentials);
    const base64Data = base64.encode(credentialsJson);

    console.log('Sending WiFi credentials');
    await this.connectedDevice.writeCharacteristicWithResponseForService(
      BLE_CONSTANTS.SERVICE_UUID,
      BLE_CONSTANTS.WIFI_CHARACTERISTIC_UUID,
      base64Data
    );
  }

  // Disconnect from device
  async disconnect(): Promise<void> {
    if (this.connectedDevice) {
      console.log('Disconnecting from device');
      await this.connectedDevice.cancelConnection();
      this.connectedDevice = null;
    }
  }

  // Cleanup
  destroy(): void {
    this.stopScan();
    this.disconnect();
    this.manager.destroy();
  }
}

// Export singleton instance
const BLEService = new BLEServiceClass();
export default BLEService;