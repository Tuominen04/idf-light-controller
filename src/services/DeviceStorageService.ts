import AsyncStorage from '@react-native-async-storage/async-storage';

const DEVICES_STORAGE_KEY = '@saved_devices';

export interface SavedDevice {
  id: string;
  name: string;
  ip: string;
  version: string;
  lastConnected: string;
}

class DeviceStorageService {
  // Get all saved devices
  async getDevices(): Promise<SavedDevice[]> {
    try {
      const devicesJson = await AsyncStorage.getItem(DEVICES_STORAGE_KEY);
      return devicesJson ? JSON.parse(devicesJson) : [];
    } catch (error) {
      console.error('Failed to get devices:', error);
      return [];
    }
  }

  // Save or update a device
  async saveDevice(deviceInfo: Omit<SavedDevice, 'lastConnected'>): Promise<boolean> {
    try {
      const devices = await this.getDevices();
      const existingIndex = devices.findIndex(d => d.id === deviceInfo.id);
      
      const deviceToSave: SavedDevice = {
        ...deviceInfo,
        lastConnected: new Date().toISOString(),
      };

      if (existingIndex >= 0) {
        // Update existing device
        devices[existingIndex] = deviceToSave;
      } else {
        // Add new device
        devices.push(deviceToSave);
      }

      await AsyncStorage.setItem(DEVICES_STORAGE_KEY, JSON.stringify(devices));
      return true;
    } catch (error) {
      console.error('Failed to save device:', error);
      return false;
    }
  }

  // Delete a device
  async deleteDevice(deviceId: string): Promise<boolean> {
    try {
      const devices = await this.getDevices();
      const filteredDevices = devices.filter(d => d.id !== deviceId);
      await AsyncStorage.setItem(DEVICES_STORAGE_KEY, JSON.stringify(filteredDevices));
      return true;
    } catch (error) {
      console.error('Failed to delete device:', error);
      return false;
    }
  }

  // Get a specific device by ID
  async getDevice(deviceId: string): Promise<SavedDevice | null> {
    try {
      const devices = await this.getDevices();
      return devices.find(d => d.id === deviceId) || null;
    } catch (error) {
      console.error('Failed to get device:', error);
      return null;
    }
  }

  // Clear all devices
  async clearAllDevices(): Promise<boolean> {
    try {
      await AsyncStorage.removeItem(DEVICES_STORAGE_KEY);
      return true;
    } catch (error) {
      console.error('Failed to clear devices:', error);
      return false;
    }
  }

  // Check if a device exists
  async deviceExists(deviceId: string): Promise<boolean> {
    const devices = await this.getDevices();
    return devices.some(d => d.id === deviceId);
  }
}

export default new DeviceStorageService();