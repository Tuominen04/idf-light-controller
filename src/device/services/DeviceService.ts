import { Alert } from "react-native";
import HTTPService, { OTAProgress } from "../../http/HTTPService";
import DeviceStorageService, { SavedDevice } from "./DeviceStorageService";
import { otaUrl  } from "../../credentials";

async function getOtaProgress(
  savedDevice: SavedDevice, 
  setOTAProgress: (OTAProgress: OTAProgress) => void, 
  setOtaLoading: (loading: boolean) => void,
) {
    try {
      const httpDevice = await DeviceStorageService.getDevice(savedDevice.id);
      const progress = await HTTPService.getOTAProgress(savedDevice.ip);

      const otaProgress =  { 
        status: progress.status,
        in_progress: !!httpDevice?.otaStatus || progress.in_progress, 
        progress: progress.progress 
      };

      setOTAProgress(otaProgress);
      if (otaProgress?.in_progress) {
        setOtaLoading(true);
        setOtaLoading(false);
      }
    } catch (error) {
      console.warn('Failed to get OTA progress:', error);
  }
}

async function savedInfo(
  device: SavedDevice, 
  setIsConnected: (connected: boolean) => void,
  setOTAProgress: (OTAProgress: OTAProgress) => void, 
  setOtaLoading: (loading: boolean) => void,
) {
  try {
    const savedDevice = await DeviceStorageService.getDevice(device.id);
    if (savedDevice) {
      getOtaProgress(savedDevice, setOTAProgress, setOtaLoading);
      setIsConnected(true);
    } else {
      console.warn('No saved device info found');
    }
  } catch (error) {
    console.error('Failed to get saved device info:', error);
  }
}

export async function deviceInfo(
  device: SavedDevice, 
  setLoading: (loading: boolean) => void, 
  setIsConnected: (connected: boolean) => void, 
  setLightState: (state: 'on' | 'off') => void
) {
  try {
    const online = await HTTPService.checkConnection(device.ip);
    setIsConnected(online);
    if (!online) {
      console.error('Device is not connected');
      return;
    }

    // Get light status
    const status = await HTTPService.getLightStatus(device.ip);
    if (!status) {
      throw new Error('No light status received');
    }
    setLightState(status.state);

    const info = await HTTPService.getFirmwareInfo(device.ip);
    if (!info) {
      throw new Error('No firmware info received');
    }

    // Create a SavedDevice object and save it
    const prev = await DeviceStorageService.getDevice(device.id);
    const savedDevice: SavedDevice = {
      ...prev, // keep previous fields like projectName, buildDate, otaStatus, etc.
      id: device.id,
      name: device.name,
      ip: device.ip,
      version: info.version,
      lastConnected: new Date().toISOString(),
    };
    await DeviceStorageService.saveDevice(savedDevice);

  } catch (error) {
    console.error('Failed to get firmware info:', error);
  } finally {
    setLoading(false);
  }
}

export async function deviceConnection(
  device: SavedDevice, 
  setIsConnected: (connected: boolean) => void,
  setLoading: (loading: boolean) => void,
  setLightState: (state: 'on' | 'off') => void,
  setOTAProgress: (OTAProgress: OTAProgress) => void, 
  setOtaLoading: (loading: boolean) => void,
){
    try {
      await savedInfo(device, setIsConnected, setOTAProgress, setOtaLoading);
      await deviceInfo(device, setLoading, setIsConnected, setLightState);
    } catch (error) {
      console.error('Connection check failed:', error);
      setIsConnected(false);
    }
  };

export async function toggleLight(
  isConnected: boolean, 
  device: SavedDevice, 
  setLoading: (loading: boolean) => void, 
  setLightState: (state: 'on' | 'off') => void
) {
  if (!isConnected) {
    Alert.alert('Device Offline', 'Device is not connected. Please check your network connection.');
    return;
  }

  setLoading(true);
  try {
    const newStateText = await HTTPService.toggleLight(device.ip);
    setLightState(newStateText.toLowerCase() as 'on' | 'off');
  } catch (error) {
    console.error('Toggle failed:', error);
    Alert.alert('Error', 'Failed to toggle light. Please try again.');
  } finally {
    setLoading(false);
  }
}

export async function startOTAUpdate(
  device: SavedDevice,
  setLoading: (loading: boolean) => void, 
  setShowOTAModal: (show: boolean) => void,
) {
  setLoading(true);
  try {
    await HTTPService.startOTAUpdate(device.ip, otaUrl);
    setShowOTAModal(false);
    
    // Start monitoring OTA progress
    const deviceInfo = await DeviceStorageService.getDevice(device.id);
    if (!deviceInfo) {
      console.error('Device not found in storage');
      return;
    }
    const deviceToSave: SavedDevice = {
      ...deviceInfo,
      otaStatus: true, // Mark OTA as in progress
    };
    await DeviceStorageService.saveDevice(deviceToSave);
  } catch (error) {
    console.error('OTA update failed:', error);
    Alert.alert('Error', 'Failed to start OTA update. Please check the URL and try again.');
  } finally {
    setLoading(false);
  }
}

export function confirmDelete(device: SavedDevice, navigation: any) {
  Alert.alert(
    'Delete Device',
    `Are you sure you want to delete ${device.name}?`,
    [
      { text: 'Cancel', style: 'cancel' },
      { text: 'Delete', onPress: () => {
        DeviceStorageService.deleteDevice(device.id).then(success => {
          if (success) {
            navigation.navigate('Home');
          } else {
            Alert.alert('Error', 'Failed to delete device. Please try again.');
          }
        });
      }, style: 'destructive' },
    ]
  );
}

