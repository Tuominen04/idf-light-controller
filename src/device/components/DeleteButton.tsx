import { Text, TouchableOpacity } from "react-native";
import styles from "../../styles/DeviceControlScreen.styles";
import { confirmDelete } from "../services/DeviceService";
import { SavedDevice } from "../DeviceStorageService";

export const renderDeleteButton = (loading: boolean, isConnected: boolean, device: SavedDevice, navigation: any) => (
    <TouchableOpacity
      style={[styles.deleteButton, { 
        opacity: loading || !isConnected ? 0.5 : 1 
      }]}
      onPress={() => confirmDelete(device, navigation)}
      disabled={loading || !isConnected}
    >
      <Text style={styles.deleteButtonText}>Delete Device</Text>
    </TouchableOpacity>
  );
