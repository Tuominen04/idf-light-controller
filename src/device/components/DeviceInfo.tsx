/**
 * @file DeviceInfo.tsx
 * @description
 * UI helpers for rendering device connection status and device information card.
 * 
 * @remarks
 * - Used in DeviceControlScreen and related components.
 * - Relies on styles from DeviceControlScreen.styles.
 * - Uses SavedDevice type and formatDateTime utility.
 */

import { Text, View } from "react-native";
import styles from "../../styles/DeviceControlScreen.styles";
import { SavedDevice } from "../services/DeviceStorageService";
import { formatDateTime } from "../../utils/FormatDateTime";

/**
 * Renders the connection status indicator for a device.
 * 
 * @param isConnected - Whether the device is currently connected.
 * @returns A styled view showing a colored dot and status text ("Connected" or "Offline").
 */
export const renderConnectionStatus = (isConnected: boolean) => (
    <View style={styles.statusContainer}>
      <View style={[styles.statusDot, { backgroundColor: isConnected ? '#4CAF50' : '#f44336' }]} />
      <Text style={[styles.statusText, { color: isConnected ? '#4CAF50' : '#f44336' }]}>
        {isConnected ? 'Connected' : 'Offline'}
      </Text>
    </View>
  );

/**
 * Renders a card displaying detailed information about a device.
 * 
 * @param device - The SavedDevice object containing device details.
 * @returns A styled view with device fields and values.
 */
export const renderDeviceInfo = (device: SavedDevice) => (
    <View style={styles.controlCard}>
      <Text style={styles.cardTitle}>Device Information</Text>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>Name:</Text>
        <Text style={styles.infoValue}>{device.name}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>ID:</Text>
        <Text style={styles.infoValue}>{device.id}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>IP Address:</Text>
        <Text style={styles.infoValue}>{device.ip}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>Project:</Text>
        <Text style={styles.infoValue}>{device.projectName}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>Version:</Text>
        <Text style={styles.infoValue}>{device.version}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>Last Connected:</Text>
        <Text style={styles.infoValue}>{formatDateTime(device.lastConnected)}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>Build Date:</Text>
        <Text style={styles.infoValue}>{formatDateTime(device.buildDate ?? '')}</Text>
      </View>
      <View style={styles.infoRow}>
        <Text style={styles.infoLabel}>OTA Status:</Text>
        <Text style={[styles.infoValue, { color: device.otaStatus ? '#FF9800' : '#4CAF50' }]}>
          {device.otaStatus ? 'In Progress' : 'Ready'}
        </Text>
      </View>
    </View>
  );