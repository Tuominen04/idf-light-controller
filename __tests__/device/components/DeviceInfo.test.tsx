/**
 * @file DeviceInfo.test.tsx
 * @description
 * Unit tests for DeviceInfo component helpers: renderConnectionStatus and renderDeviceInfo.
 * 
 * @remarks
 * - Verifies correct rendering of connection status and device information.
 * - Includes snapshot and style tests for comprehensive coverage.
 * 
 * @note
 * - If the UI changes intentionally, update the snapshot by running `npm test -- -u` (on Windows) and commit the new snapshot.
 */

import { render, screen, waitFor } from "@testing-library/react-native";
import { renderConnectionStatus, renderDeviceInfo } from '../../../src/device/components/DeviceInfo';
import { SavedDevice } from "../../../src/device/services/DeviceStorageService";
import timezone_mock from 'timezone-mock';

timezone_mock.register('UTC');

// Tests for the renderConnectionStatus helper
describe('DeviceInfo renderConnectionStatus', () => {
  /**
   * Should show "Connected" status when isConnected is true.
   */
  test('shows connected status when isConnected is true', async () => {
    render(renderConnectionStatus(true));

    await waitFor(() => {
      expect(screen.getByText(/connected/i)).toBeTruthy();
    });
  });

  /**
   * Should show "Offline" status when isConnected is false.
   */
  test('shows offline status when isConnected is false', async () => {
    render(renderConnectionStatus(false));

    await waitFor(() => {
      expect(screen.getByText(/offline/i)).toBeTruthy();
    });
  });
});

// Tests for the renderDeviceInfo helper
describe('DeviceInfo renderDeviceInfo', () => {
  // Default device mock for tests
  const defaultDevice: SavedDevice = {
    id: '1',
    name: 'Test Device',
    ip: '192.168.1.1',
    projectName: 'Test Project',
    version: '1.0.0',
    lastConnected: '2025-01-01T00:00:00Z',
    buildDate: '2025-02-02T00:00:00Z',
    otaStatus: false
  };

  /**
   * Should render the device information card even with empty device data.
   */
  test('renders device information screen', async () => {
    render(renderDeviceInfo({} as SavedDevice));

    await waitFor(() => {
      expect(screen.getByText(/Device Information/i)).toBeTruthy();
    });
  });

  /**
   * Should render all device information fields and values correctly.
   */
  test('renders device information correctly', async () => {
    render(renderDeviceInfo(defaultDevice));

    await waitFor(() => {
      expect(screen.getByText(/Name:/i)).toBeTruthy();
      expect(screen.getByText(/Test Device/i)).toBeTruthy();
      expect(screen.getByText(/ID:/i)).toBeTruthy();
      expect(screen.getByText(/^1$/i)).toBeTruthy();
      expect(screen.getByText(/IP Address:/i)).toBeTruthy();
      expect(screen.getByText(/192.168.1.1/i)).toBeTruthy();
      expect(screen.getByText(/Project:/i)).toBeTruthy();
      expect(screen.getByText(/Test Project/i)).toBeTruthy();
      expect(screen.getByText(/Version:/i)).toBeTruthy();
      expect(screen.getByText(/1.0.0/i)).toBeTruthy();
      expect(screen.getByText(/Last Connected:/i)).toBeTruthy();
      expect(screen.getByText(/2025-01-01 00:00/i)).toBeTruthy();
      expect(screen.getByText(/Build Date:/i)).toBeTruthy();
      expect(screen.getByText(/2025-02-02 00:00/i)).toBeTruthy();
      expect(screen.getByText(/OTA Status:/i)).toBeTruthy();
      expect(screen.getByText(/Ready/i)).toBeTruthy();
    });
  });

  /**
   * Snapshot test for the device info card.
   * Ensures the rendered output matches the saved snapshot.
   */
  test('matches snapshot for device info', () => {
    const device: SavedDevice = {
      id: '1',
      name: 'Test Device',
      ip: '192.168.1.1',
      projectName: 'Test Project',
      version: '1.0.0',
      lastConnected: '2025-01-01T00:00:00Z',
      buildDate: '2025-02-02T00:00:00Z',
      otaStatus: false
    };
    const tree = render(renderDeviceInfo(device)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  /**
   * Should render "In Progress" when otaStatus is true.
   */
  test('renders "In Progress" when otaStatus is true', () => {
    const device = { ...defaultDevice, otaStatus: true };
    render(renderDeviceInfo(device));
    expect(screen.getByText(/In Progress/i)).toBeTruthy();
  });

  /**
   * Should render correct color style for OTA status when in progress.
   */
  test('renders correct color for OTA status', () => {
    const device = { ...defaultDevice, otaStatus: true };
    render(renderDeviceInfo(device));
    const otaText = screen.getByText(/In Progress/i);
    expect(otaText.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ color: '#FF9800' })
      ])
    );
  });
});