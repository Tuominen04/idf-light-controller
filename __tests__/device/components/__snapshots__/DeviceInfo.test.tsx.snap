// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceInfo renderDeviceInfo matches snapshot for device info 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderRadius": 12,
      "elevation": 3,
      "margin": 16,
      "padding": 20,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 20,
        "fontWeight": "600",
        "marginBottom": 16,
      }
    }
  >
    Device Information
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      Name:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      Test Device
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      ID:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      1
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      IP Address:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      192.168.1.1
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      Project:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      Test Project
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      Version:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      1.0.0
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      Last Connected:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      2025-01-01 02:00
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      Build Date:
    </Text>
    <Text
      style={
        {
          "color": "#333",
          "flex": 2,
          "fontSize": 14,
          "textAlign": "right",
        }
      }
    >
      2025-02-02 02:00
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666",
          "flex": 1,
          "fontSize": 14,
        }
      }
    >
      OTA Status:
    </Text>
    <Text
      style={
        [
          {
            "color": "#333",
            "flex": 2,
            "fontSize": 14,
            "textAlign": "right",
          },
          {
            "color": "#4CAF50",
          },
        ]
      }
    >
      Ready
    </Text>
  </View>
</View>
`;
